### ISSUE TITLE
Review and revise exception handling within Action Scheduler

### ISSUE THREAD

**Issue by alewolf** [Prospress/action-scheduler#1059]
There are quite a few occasions when Action Scheduler experiences an unexpected condition, and (mostly for safety) it stops. It communicates about this using exceptions, but many of these exceptions:

- Are not handled by Action Scheduler.
- The design does not make it easy for parent plugins to handle them.

In most cases, this is 'fine' in the sense that the queue is processed in requests that are independent of user-facing requests. However, they could impact other scheduled events when running in the context of WP Cron.

Additionally, there is the problem noted below which is that only one version of Action Scheduler can be active at any one time. That is, only one parent plugin 'wins' this particular race, and if things go wrong—irrespective of the underlying reason—the backtraces often lead users back to that plugin's support team.

Let's review and revise when we throw exceptions, and when we should add handling for them.

- We could consider logging deadlocks etc in Action Scheduler's own logging table (perhaps by referencing an `action_id` of `0` to indicate they are not related to a specific action). If we do this, we would also want to add:
    - A means of viewing any such global log entries.
    - Possibly, hooks that fire when these conditions are encountered.
- There are other cases where we bail out for safety, but possibly are being 'too safe'. For instance, when cleaning up expired actions: if the action being removed is concurrently removed by another process, we currently bail out of the clean-up process. This is probably overkill.
- Decisions like the above could also be made configurable/overrideable.

Original report follows...

---

I bundle the ActionScheduler version `3.7.4` in the [Pixel Manager for WooCommerce](http://wordpress.org/plugins/woocommerce-google-adwords-conversion-tracking-tag/). 

As of recently we started to receive critical error reports like this: 

```
2024-05-13T08:24:09+00:00 Critical Uncaught InvalidArgumentException: Unidentified action 19061390 in /nas/content/live/nck/wp-content/plugins/pixel-manager-pro-for-woocommerce/vendor/woocommerce/action-scheduler/classes/data-stores/ActionScheduler_DBStore.php:1086  
Additional context
{
    "error": {
        "type": 1,
        "file": "\/nas\/content\/live\/nck\/wp-content\/plugins\/pixel-manager-pro-for-woocommerce\/vendor\/woocommerce\/action-scheduler\/classes\/data-stores\/ActionScheduler_DBStore.php",
        "line": 1086
    },
    "backtrace": [
        "",
        "#0 \/nas\/content\/live\/nck\/wp-content\/plugins\/pixel-manager-pro-for-woocommerce\/vendor\/woocommerce\/action-scheduler\/classes\/abstracts\/ActionScheduler_Abstract_QueueRunner.php(120): ActionScheduler_DBStore->mark_failure(19061390)",
        "#1 \/nas\/content\/live\/nck\/wp-content\/plugins\/pixel-manager-pro-for-woocommerce\/vendor\/woocommerce\/action-scheduler\/classes\/abstracts\/ActionScheduler_Abstract_QueueRunner.php(98): ActionScheduler_Abstract_QueueRunner->handle_action_error(19061390, Object(Exception), 'Async Request', true)",
        "#2 \/nas\/content\/live\/nck\/wp-content\/plugins\/pixel-manager-pro-for-woocommerce\/vendor\/woocommerce\/action-scheduler\/classes\/ActionScheduler_QueueRunner.php(169): ActionScheduler_Abstract_QueueRunner->process_action(19061390, 'Async Request')",
        "#3 \/nas\/content\/live\/nck\/wp-content\/plugins\/pixel-manager-pro-for-woocommerce\/vendor\/woocommerce\/action-scheduler\/classes\/ActionScheduler_QueueRunner.php(139): ActionScheduler_QueueRunner->do_batch(25, 'Async Request')",
        "#4 \/nas\/content\/live\/nck\/wp-includes\/class-wp-hook.php(324): ActionScheduler_QueueRunner->run('Async Request')",
        "#5 \/nas\/content\/live\/nck\/wp-includes\/class-wp-hook.php(348): WP_Hook->apply_filters('', Array)",
        "#6 \/nas\/content\/live\/nck\/wp-includes\/plugin.php(517): WP_Hook->do_action(Array)",
        "#7 \/nas\/content\/live\/nck\/wp-content\/plugins\/pixel-manager-pro-for-woocommerce\/vendor\/woocommerce\/action-scheduler\/classes\/ActionScheduler_AsyncRequest_QueueRunner.php(52): do_action('action_schedule...', 'Async Request')",
        "#8 \/nas\/content\/live\/nck\/wp-content\/plugins\/woocommerce\/includes\/libraries\/wp-async-request.php(147): ActionScheduler_AsyncRequest_QueueRunner->handle()",
        "#9 \/nas\/content\/live\/nck\/wp-includes\/class-wp-hook.php(324): WP_Async_Request->maybe_handle('')",
        "#10 \/nas\/content\/live\/nck\/wp-includes\/class-wp-hook.php(348): WP_Hook->apply_filters('', Array)",
        "#11 \/nas\/content\/live\/nck\/wp-includes\/plugin.php(517): WP_Hook->do_action(Array)",
        "#12 \/nas\/content\/live\/nck\/wp-admin\/admin-ajax.php(207): do_action('wp_ajax_nopriv_...')",
        "#13 {main}",
        "thrown"
    ]
}
```

1. In one particular case that I analysed the scheduled action didn't even come from the Pixel Manager. But, WooCommerce and the Pixel Manager both bundle version `3.7.4` of the ActionScheduler. It seems that the version shipped by the Pixel Manager is loaded first (probably because of the alphabetical order). The error path shows the Pixel Manager and we get all support requests. 
2. There is no helpful information in that error. In the case I analysed the action `19061390` exists and has just failed to execute. 
3. What's even more annoying is that one of our customers reached out to WooCommerce support and they reverted him to us, because the error was thrown from the ActionScheduler that's shipped in the Pixel Manager. Even though the action probably was not even scheduled by the Pixel Manager. Since WooCommerce **is** the maintainer of the ActionScheduler you should also support users who have issues with it. 
4. I will now revert back to a lower version of the ActionScheduler, just to avoid further support requests. But, you should really do something about this problem. All other plugins that ship version `3.7.4` will have the same issue. Unwarranted support requests because the ActionScheduler doesn't handle failed actions gracefully and logs them in the **fatal-errors** log.

**Comment by barryhughes** [comment]
Hi @alewolf,

Are you quite sure the problem is specifically tied to [3.7.4](https://github.com/woocommerce/action-scheduler/releases/tag/3.7.4)? This was a fairly [minimal set of changes](https://github.com/woocommerce/action-scheduler/compare/3.7.3...3.7.4). What version did you switch to, and have you found that that has resolved the problems (at least in cases where WooCommerce, or other plugins shipping 3.7.4, are not present)?

> There is no helpful information in that error.

We could probably finesse some of these, thank you for flagging. For what it's worth, errors of this type may point to a race condition between two queue runners, or between Action Scheduler's own queue runners and some custom code.

> What's even more annoying is that one of our customers reached out to WooCommerce support and they reverted him to us, because the error was thrown from the ActionScheduler that's shipped in the Pixel Manager. Even though the action probably was not even scheduled by the Pixel Manager. Since WooCommerce is the maintainer of the ActionScheduler you should also support users who have issues with it.

I'm sure this can be annoying, but things are not always quite as clear cut as you are suggesting. We *do* support Action Scheduler, here and to a lesser extent in the WordPress community forums, but Action Scheduler is a conduit for processing work: errors surface *through* it, but not always *because* of it.

With your experience, you may understand when and how to make the distinction, but I'm not sure it's realistic to expect large support teams to make the right call 100% of the time. Even in this case, it's not completely clear that the problem has its origin in Action Scheduler.

> I will now revert back to a lower version of the ActionScheduler, just to avoid further support requests. But, you should really do something about this problem. All other plugins that ship version 3.7.4 will have the same issue. 

These exceptions are not new, and as above I'm not convinced they are tied to 3.7.4 specifically. Definitely let me know if I'm missing something obvious, here.

> Unwarranted support requests because the ActionScheduler doesn't handle failed actions gracefully and logs them in the fatal-errors log.

Though I can take some guesses as to what you'd like to see here, I'd much rather get your informed opinion. When an unexpected condition is met, should it: fail silently, log the errors somewhere else, carry on regardless? When should we let site operators know of problems, and when should we suppress them?

**Comment by alewolf** [comment]
Hi @barryhughes 

> Are you quite sure the problem is specifically tied to [3.7.4](https://github.com/woocommerce/action-scheduler/releases/tag/3.7.4)? This was a fairly [minimal set of changes](https://github.com/woocommerce/action-scheduler/compare/3.7.3...3.7.4). What version did you switch to, and have you found that that has resolved the problems (at least in cases where WooCommerce, or other plugins shipping 3.7.4, are not present)?

I upgraded from `3.7.2` to `3.7.4`, then the support requests started. 

I can't say if it is bound to `3.7.4` or which version exactly. I only can say when upgrading to the same version as what's bundled in WooCommerce, then my bundled version will take precedence. And whatever code triggers those errors, users will contact **our** support because the ActionScheduler loads from our bundle and the Pixel Manager is visible in the path. 

> These exceptions are not new, and as above I'm not convinced they are tied to 3.7.4 specifically. Definitely let me know if I'm missing something obvious, here.

As explained in my earlier post, I've downgraded to one version below WooCommerce. As expected it will now hit another install. Now Rank Math will get those support requests: 

```
Uncaught InvalidArgumentException: Unidentified action 19146968 in /nas/content/live/nck/wp-content/plugins/seo-by-rank-math/vendor/woocommerce/action-scheduler/classes/data-stores/ActionScheduler_DBStore.php:1086
```

> Though I can take some guesses as to what you'd like to see here, I'd much rather get your informed opinion. When an unexpected condition is met, should it: fail silently, log the errors somewhere else, carry on regardless? When should we let site operators know of problems, and when should we suppress them?

* Add more information to the exception message. I only can say that there are more failed ActionScheduler tasks, but the exception triggers only on a subset. I looked into the database and compared a few. I couldn't deduct a pattern that would explain why the exception triggers on some. 

* If WooCommerce is active you could log it into a separate source, not into the fatal-error log. 

* I dont' think it should fail silently. You need the telemetry to be able to find what's going on. 

* When more than one plugin is loading the same version of the ActionScheduler as the one bundled in WooCommerce, then the one bundled with WooCommerce should take precedence.

**Comment by barryhughes** [comment]
> - When more than one plugin is loading the same version of the ActionScheduler as the one bundled in WooCommerce, then the one bundled with WooCommerce should take precedence.

This doesn't feel like the right direction, to me. 

I'm curious, though: in the case of your plugin it seems like [WooCommerce is a requirement](https://plugins.trac.wordpress.org/browser/woocommerce-google-adwords-conversion-tracking-tag/tags/1.43.0/wgact.php#L273). What prevents you from using the WC Queue, or from using WooCommerce's copy of Action Scheduler directly (vs bundling your own)?

**Comment by alewolf** [comment]
> What prevents you from using the WC Queue, or from using WooCommerce's copy of Action Scheduler directly (vs bundling your own)?

Good question. 

Technically the Pixel Manager can also be used on installs where WooCommerce is not running (because some of our users don't want to use more than one tagging solution). 

Until now I haven't thought about loading the Action Scheduler conditionally, depending if WooCommerce is active or not. 

However, I fear that it might not be possible to implement conditional loading. 

The Action Scheduler needs to be required very early in the loading process. If I understand correctly this is to make sure that the Action Scheduler can check if only one, and only the most recent version is loaded, in case more than one plugin tries to load it. However, it is not possible to check if WooCommerce is active so early. 

Let me know if you think this **is** possible. If so, I will change the logic and only load it conditionally and let WooCommerce take over.

**Comment by barryhughes** [comment]
> Technically the Pixel Manager can also be used on installs where WooCommerce is not running.

Gotcha, I misunderstood in that case.

> Let me know if you think this is possible.

It probably is *possible,* but given the context I suspect the potential risk and complexity may not be worth the gain.

**Comment by barryhughes** [comment]
> Occasionally seeing a failed action with this message:
>
> > `action failed via WP Cron: Unable to update the status of action ###### to in-progress.`
>
> Did some Googling etc but not able to find any real explanation or solution.

This was reported on the [community forums](https://wordpress.org/support/topic/unable-to-update-the-status-of-action-to-in-progress/). Noting it here because, similar to the work we did in https://github.com/woocommerce/action-scheduler/pull/1060, we can probably add additional context to make things clearer.

**Comment by barryhughes** [comment]
Updated title and description. We're hoping to look at options here soon.

**Comment by crstauf** [comment]
@barryhughes Would this error be one that you would expect an author of a plugin (bundling AS) to handle?

```
Uncaught InvalidArgumentException: Unidentified action 19146968
```

That is to say, if a customer came to WooCommerce support about the above error, would your internal process direct to investigate further, or simply send away to whichever plugin bundled Action Scheduler?

Certainly there are errors that are because of Action Scheduler, and not because of the bundling plugin. Perhaps including in the error message direction on who is the best first contact would be helpful. Off the top of my head, this could potentially be determined by checking `$wp_filter[ $hook ]` for the callbacks on the scheduled action.

I think @alewolf has a very valid point, that plugin support teams can potentially get flooded with support requests that are not related to their plugin, and everyone involved should be do all they can to keep that from happening, especially Action Scheduler, being the common technology. If it's not handled well, then plugin authors will do exactly as @alewolf has done, and use older versions in an attempt to avoid the support burden. Not exactly a winning approach to encourage developers to use the latest version.

**Comment by crstauf** [comment]
@alewolf In terms of conditionally loading, the [introduction of Plugin Dependencies in WP 6.5](https://make.wordpress.org/core/2024/03/05/introducing-plugin-dependencies-in-wordpress-6-5/) may give you a more viable approach:

```php
/**
 * Plugin name: Pixel Manager for WooCommerce
 * Requires Plugins: action-scheduler
 */

...

if ( version_compare( $wp_version, '6.5' ) < 0) {
    // require bundled AS
}
```

This does add an additional step for site admins, but is an option to use the latest version while also alleviating support tickets. As a site admin, I always prefer installing AS explicitly anyways; maybe there are more like me.

**Comment by barryhughes** [comment]
> Certainly there are errors that are because of Action Scheduler

Definitely. Others are not, and yet others occupy a grey area somewhere in-between (and relate to the volume of work being processed, alongside the available resources).

> I think @alewolf has a very valid point.

Yep, so do we, hence the latest update.

**Comment by crstauf** [comment]
> Yep, so do we, hence the latest update.

Oh sorry, I thought this ticket was generalized to changing exception handling and messages, not specific to alleviating false support tickets.

**Comment by barryhughes** [comment]
Yes—that's the primary technical focus. The support aspect is one of the motivations for doing this, and something we'll look to alleviate through those changes.

**Comment by barryhughes** [comment]
```
ActionScheduler_ActionFactory::repeat()
    classes/ActionScheduler_ActionFactory.php:233
```

↑ This is already caught, by `ActionScheduler_Abstract_QueueRunner::schedule_next_instance()`, which is the only location where we call it. Third parties invoking this directly are responsible for their own exception handling 🟢

**Comment by barryhughes** [comment]
```
ActionScheduler_Store::validate_args()
    classes/abstracts/ActionScheduler_Store.php:324
```

↑ This is already caught, by `ActionScheduler_DBStore::fetch_action()`, which is the only location where we call it. Third parties invoking this directly are responsible for their own exception handling 🟢 

(Edit: 2 exceptions are thrown from this method, but both are covered.)

**Comment by barryhughes** [comment]
```
ActionScheduler_Store::validate_action()
    classes/abstracts/ActionScheduler_Store.php:359
```

↑ This is already caught, by `ActionScheduler_DBStore::save_action_to_db()` Third parties invoking this directly are responsible for their own exception handling 🟢

**Comment by barryhughes** [comment]
```
ActionScheduler_Abstract_QueueRunner::process_action()
    classes/abstracts/ActionScheduler_Abstract_QueueRunner.php:64
```

```
ActionScheduler_Abstract_QueueRunner::process_action()
    classes/abstracts/ActionScheduler_Abstract_QueueRunner.php:95
```

↑ Both of the above are part of the current nested try/catch used for error handling. The exceptions are caught in the very same structure, in the same method (but we can probably simplify this now that we don't need to support PHP 5.6) 🟠 

Let's review the nested try/catch. Low priority.

**Comment by barryhughes** [comment]
```
ActionScheduler_Abstract_ListTable::get_bulk_actions()
    classes/abstracts/ActionScheduler_Abstract_ListTable.php:162
```

↑ Unclear who/what we're guarding against here. I'd think if a bulk action lacks a valid callback, we should log that but simply omit it from the returned array of possible actions.

Needs attention, low priority (I don't recall seeing reports about this specific exception) 🟠

**Comment by barryhughes** [comment]
```
ActionScheduler_WPCLI_QueueRunner::__construct()
    classes/WP_CLI/ActionScheduler_WPCLI_QueueRunner.php:33
```

↑ This code should only run in the context of WP CLI. The exception seems justified, no follow-up needed at present 🟢

**Comment by barryhughes** [comment]
```
ProgressBar::__construct()
    classes/WP_CLI/ProgressBar.php:47
```

↑ This code should only run in the context of WP CLI. The exception seems justified, no follow-up needed at present 🟢

**Comment by barryhughes** [comment]
```
ActionScheduler_DBStore::save_action_to_db()
    classes/data-stores/ActionScheduler_DBStore.php:113
```

```
ActionScheduler_DBStore::save_action_to_db()
    classes/data-stores/ActionScheduler_DBStore.php:118
```

```
ActionScheduler_DBStore::save_action_to_db()
    classes/data-stores/ActionScheduler_DBStore.php:126
```

↑ For all of the above, the exception is thrown from within a private method but we trigger this through various paths. In some cases, we [catch the exception](https://github.com/woocommerce/action-scheduler/blob/3.8.1/classes/ActionScheduler_ActionFactory.php#L318-L329) then log an error. In others, [we do not](https://github.com/woocommerce/action-scheduler/blob/3.8.1/classes/ActionScheduler_ActionFactory.php#L193-L202) and execution probably halts. Needs review 🔴

**Comment by barryhughes** [comment]
```
ActionScheduler_DBStore::get_query_actions_sql()
    classes/data-stores/ActionScheduler_DBStore.php:397
```

```
ActionScheduler_DBStore::get_query_actions_sql()
    classes/data-stores/ActionScheduler_DBStore.php:455
```

```
ActionScheduler_DBStore::get_query_actions_sql()
    classes/data-stores/ActionScheduler_DBStore.php:470
```

```
ActionScheduler_DBStore::get_query_actions_sql()
    classes/data-stores/ActionScheduler_DBStore.php:493
```

↑ These are all within `ActionScheduler_DBStore::get_query_actions_sql()`. It's called from various public API functions like `as_get_scheduled_actions()` which lack protection against the exceptions. These particular exceptions seem somewhat reasonable, they guard against misuse that should generally be caught during development, but that can't be guaranteed. Needs review 🟠

**Comment by ws404dev** [comment]
A concrete, fully-traced instance of the problem described in this issue — including the "only one parent plugin wins the version race and gets the blame" symptom. Reproduced against **4.1.0**, so it is not fixed by upgrading.

**Affects:** 4.0.0 and **4.1.0** (latest), and the current default branch — the relevant file is byte-identical across all three.

---

## Summary

`ActionScheduler_Abstract_QueueRunner::process_action()` calls `schedule_next_instance()` **outside** its own `try/catch/finally`. The first statement of `schedule_next_instance()` is an unguarded `$this->store->get_status( $action_id )`, which throws `InvalidArgumentException` if the row is gone.

So for any **recurring** action, there is a window between `mark_complete()` and `get_status()` in which a concurrent process deleting that row turns a successfully-executed action into an **uncaught fatal that abandons every remaining action in the batch**.

This is reachable in a stock configuration: nothing exotic is required beyond (a) a recurring action and (b) any code that deletes completed actions — including code using Action Scheduler's own public `ActionScheduler::store()->delete_action()` API.

---

## Environment

| | |
|---|---|
| Action Scheduler | **4.0.0** active (bundled by WP Mail SMTP Pro 4.9.0) — **verified byte-identical to the official `4.0.0` tag** |
| Second copy present | 3.9.3 (bundled by WooCommerce) — shadowed, never loaded |
| WordPress | 6.9.7 |
| PHP | 8.3.33 |
| MySQL | 8.0.37 |
| WooCommerce | 10.8.1 |
| WooCommerce Subscriptions | 9.1.0 |
| `DISABLE_WP_CRON` | `true` (system cron hits `wp-cron.php`) |
| Persistent object cache | yes |

`ActionScheduler_Versions::latest_version()` sorts with `version_compare` and returns the **highest**, so 4.0.0 wins. Confirmed at runtime — `(new ReflectionClass('ActionScheduler_DBStore'))->getFileName()` resolves into the WP Mail SMTP Pro vendor tree. Worth noting for this issue specifically: the resulting stack traces name WP Mail SMTP Pro at every frame even though the queue being processed is WooCommerce's and Subscriptions'.

---

## The defect, at tag `4.1.0`

`classes/abstracts/ActionScheduler_Abstract_QueueRunner.php` — the outer `try` opens at line 88, its `finally` **closes at line 123**, and the reschedule call is at 125-127, outside every catch:

```php
118:        } catch ( Exception $e ) {
120:            $this->handle_action_error( $action_id, $e, $context, $valid_action );
121:        } finally {
122:            restore_error_handler();
123:        }
124:
125:        if ( isset( $action ) && is_a( $action, 'ActionScheduler_Action' ) && $action->get_schedule()->is_recurring() ) {
126:            $this->schedule_next_instance( $action, $action_id );
127:        }
```

Same file, lines 207-211 — `get_status()` is the **first** statement, is the **left** operand of the `&&` (so it always evaluates; no short-circuit protection), and is **not** inside the method's only `try` (which starts at 221 and wraps just `repeat()`):

```php
207:    protected function schedule_next_instance( ActionScheduler_Action $action, $action_id ) {
209:        if (
210:            ActionScheduler_Store::STATUS_FAILED === $this->store->get_status( $action_id )
211:            && $this->recurring_action_is_consistently_failing( $action, $action_id )
```

`classes/data-stores/ActionScheduler_DBStore.php` @ 4.1.0, lines 1425-1427:

```php
1425:        if ( is_null( $status ) ) {
1426:            throw new \InvalidArgumentException( __( 'Invalid action ID. No status found.', 'action-scheduler' ) );
```

And `delete_action()` @ 4.1.0 line 876 — the DELETE predicate is `action_id` **alone**. No `claim_id` condition, no status condition; a claimed, in-flight row is deleted just the same:

```php
876:        $deleted = $wpdb->delete( $wpdb->actionscheduler_actions, array( 'action_id' => $action_id ), array( '%d' ) );
877:        if ( empty( $deleted ) ) {
879:            throw new \InvalidArgumentException( sprintf( __( 'Unidentified action %s: we were unable to delete this action. It may may have been deleted by another process.', ... ) ) );
```

That message already anticipates exactly this class of race — but it is *post-hoc detection of a lost race*, not a guard on the `WHERE` clause.

**The fix is not in the latest release.** `ActionScheduler_Abstract_QueueRunner.php` has blob SHA `87fa61f96190171f6e4346efcc77dc19f1e363a6` on **both** tag `4.1.0` and the default branch, and sha256 `f0a54449…` on tags `4.0.0`, `4.1.0`, and on disk in production. Upgrading 4.0.0 → 4.1.0 changes nothing here.

---

## Observed failure

```
[01-Sep-2026 09:11:06 UTC] PHP Fatal error:  Uncaught InvalidArgumentException: Invalid action ID. No status found.
  in .../action-scheduler/classes/data-stores/ActionScheduler_DBStore.php:1355
Stack trace:
#0 .../ActionScheduler_Abstract_QueueRunner.php(210): ActionScheduler_DBStore->get_status(342695)
#1 .../ActionScheduler_Abstract_QueueRunner.php(126): ActionScheduler_Abstract_QueueRunner->schedule_next_instance(Object(ActionScheduler_Action), 342695)
#2 .../ActionScheduler_QueueRunner.php(219): ActionScheduler_Abstract_QueueRunner->process_action(342695, 'WP Cron')
#3 .../ActionScheduler_QueueRunner.php(187): ActionScheduler_QueueRunner->do_batch(25, 'WP Cron')
#4 .../wp-includes/class-wp-hook.php(341): ActionScheduler_QueueRunner->run('WP Cron')
#7 .../wp-cron.php(191): do_action_ref_array('action_schedule...', Array)
```

Frames #0 and #1 are precisely lines 210 and 126 above. The batch size was 25; the exception escapes `do_batch()`, so every action still unprocessed in that claim is abandoned mid-flight.

---

## Root cause

1. Runner **A** claims a batch and processes recurring action *N*. `execute()` succeeds; `mark_complete(N)` sets `status = 'complete'`.
2. Runner **B**, in a separate request, finishes its own `run()` and fires `action_scheduler_after_process_queue`. A subscriber sweeps actions by `(hook, status='complete')` and calls `ActionScheduler::store()->delete_action()` on each — which, per above, has no claim guard. Row *N* and (via `action_scheduler_deleted_action` → `ActionScheduler_DBLogger`) all of its log rows are deleted.
3. Runner **A** returns from the `finally` at line 123, evaluates `is_recurring()` → true, calls `schedule_next_instance(N)` → `get_status(N)` → row missing → **uncaught fatal**, rest of batch abandoned.

### Two details that make this much more likely than it first looks

**(a) The deleting hook fires even when the runner does no work.** In `ActionScheduler_QueueRunner::run()`, the batch loop is inside `if ( false === $this->has_maximum_concurrent_batches() ) { … }`, but `do_action( 'action_scheduler_after_process_queue' )` sits **after that block closes**. A runner that is blocked by the concurrency ceiling processes *zero* actions and still runs every `after_process_queue` subscriber. So the deleter fires on essentially every queue-runner request, not only on productive ones — which is exactly the population of requests most likely to overlap a busy runner.

**(b) Concurrent batches are enabled by a first-party plugin.** WooCommerce Subscriptions' `Queue_Management\Concurrent_Batches_Booster` filters `action_scheduler_queue_runner_concurrent_batches` from Action Scheduler's default `1` up to `2` (priority 1000) when dedicated queues are on. Verified at runtime on this site: `woocommerce_subscriptions_dedicated_queues_enabled = 'yes'`, and `apply_filters( 'action_scheduler_queue_runner_concurrent_batches', 1 )` returns **2**. Two overlapping runners are the *intended* configuration here, not a misconfiguration.

**(c) Three independent entry points start queue runners on this install**, which is what keeps two of them in flight at once:

| Entry point | Cadence | Visible in HTTP access log? |
|---|---|---|
| `GET /wp-cron.php` | every ~1-2 min | yes |
| `GET /wp-admin/admin-ajax.php?action=as_async_request_queue_runner` (loopback) | irregular, several/hour | yes |
| `wp cron event run --due-now` from system crontab | every 5 min and every 15 min | **no** |

The third is worth flagging for anyone trying to reproduce this from access logs: a WP-CLI runner leaves no HTTP trace, so an access log can look free of overlap while two runners are in fact executing. On this install the fatal fell inside a `wp-cron.php` request (log entries bracket it at 05:10:19 and 05:12:21 local), while the WP-CLI runner fires on the `*/5` schedule.

### The concrete deleter in this instance

WP Mail SMTP Pro's `ProcessQueueTask`:

- schedules `wp_mail_smtp_queue_process` via `->recurring( strtotime('now'), MINUTE_IN_SECONDS )->unique()->register()` — so `is_recurring()` is true and line 126 **is** reached;
- registers `add_action( 'action_scheduler_after_process_queue', [ $this, 'cleanup' ] )` → `remove_completed( 10 )`, which selects by `(hook, status='complete')` with **no date cutoff** and calls `ActionScheduler::store()->delete_action()` per id.

I want to be clear that I do **not** think this is primarily WP Mail SMTP's bug. It is using a documented public API (`ActionScheduler::store()->delete_action()`) on actions it owns, on a hook Action Scheduler provides for exactly this purpose. Action Scheduler currently offers no way for that call to be safe, and no way for the queue runner to survive it.

---

## Reproduction

1. Schedule a recurring action on a short interval.
2. From a second concurrent request, delete that action by id (`ActionScheduler::store()->delete_action( $id )`) in the window after the runner's `mark_complete()` and before `schedule_next_instance()`. Hooking `action_scheduler_completed_action` in a *separate process*, or simply sweeping `status='complete'` on `action_scheduler_after_process_queue`, both reach it.
3. The runner emits `Uncaught InvalidArgumentException: Invalid action ID. No status found.` and abandons the remainder of the claimed batch.

Note that a **same-request** deleter cannot trigger this: `after_process_queue` fires after the whole batch loop, and while `mark_complete()` does fire `action_scheduler_completed_action` inside the window, that is the only in-window hook. The bug genuinely requires concurrency.

---

## Supporting evidence from the affected site

All read-only. Table prefix and paths genericised.

- **The row is gone, and it went through `delete_action()` rather than raw SQL.** `action_id 342695` is absent from `actionscheduler_actions` (surviving neighbours 342693 and 342702 bracket it), *and* its `actionscheduler_logs` rows are absent — log removal is driven by `ActionScheduler_DBLogger` on `action_scheduler_deleted_action`, so log absence specifically implicates the API path. Corroborating: the two tables reconcile exactly — 3918 action rows vs 3918 `action created` log rows, and **0 orphan log rows** — so no raw-SQL deletion has occurred on this install.

- **A per-hook retention anomaly that only a hook-scoped deleter can produce.** Completed-row counts: `woocommerce_cancel_unpaid_orders` 1448, `wc_stripe_database_cache_prefetch_async` 1420, and each daily `wp_mail_smtp_*` hook exactly 31 (matching Action Scheduler's 31-day default retention; no `action_scheduler_retention_period` filter is registered anywhere on the site). Against that, `wp_mail_smtp_queue_process` has **zero** completed rows, and `wp_mail_smtp_send_enqueued_email` has no rows at all. The set of anomalous hooks is *exactly* the set whose task class registers `cleanup()` on `after_process_queue`. Every other deleter available on this install is hook-agnostic, so none of them can produce a per-hook pattern.

- **The hook does complete** (i.e. the zero above is deletion, not failure): the plugin's own email-queue table holds 424 rows, all `status = 2` (PROCESSED) with `date_processed` set, most recent `2026-09-01 08:29:01`.

- **`status='pending'` sweeps are excluded by construction.** `log_execution()` sets `STATUS_RUNNING` (`'in-progress'`) before execution and `mark_complete()` sets `'complete'` after, so a row is never `pending` anywhere in the vulnerable window.

- **Retention-based cleanup is excluded.** `ActionScheduler_QueueCleaner::clean_actions()` bounds every batch with `'modified' => $cutoff, 'modified_compare' => '<='` at 31 days (3 months for failed). A row completed milliseconds ago has `modified = now` and can never match.

---

## Suggested fixes

In rough order of value:

1. **Make the reschedule safe.** Either move the `is_recurring()` / `schedule_next_instance()` block inside the existing `try/catch`, or have `schedule_next_instance()` tolerate a missing row — e.g. treat "row gone" as "the action was deliberately removed; do not reschedule, do not throw," optionally with a `do_action` so callers can observe it. A row that vanished between completion and reschedule is a legitimate state (something cancelled or cleaned it), not a programming error.

2. **Don't let `get_status()` be the vehicle for a control-flow decision it can't safely make.** `schedule_next_instance()` only needs to know *whether the status is `failed`*. A nullable accessor (`get_status_or_null()`), or simply catching around line 210, removes the throw from this path entirely without changing `get_status()`'s contract for other callers.

3. **Disambiguate `get_status()`'s null.** `$wpdb->get_var()` returns `null` both when the row is missing **and when the query fails**. This install logs 22 `Commands out of sync` wpdb errors — 14 against `actionscheduler_*` tables, with bursts at 09:00:56 and 09:01:00 on the same morning as this fatal. (For *this* event that alternative is excluded: no wpdb error is logged at 09:11:06, and the row plus its logs are gone.) But in general `InvalidArgumentException: Invalid action ID` is a misleading diagnosis for a transient DB failure, and it sends people hunting for a data-integrity problem that isn't there. Checking `$wpdb->last_error` before concluding "no such row" would help a lot of bug reports.

4. **Consider guarding `delete_action()`,** or documenting the hazard. Deleting a row another runner holds a claim on is currently silent and unconditional. At minimum, `remove_completed`-style cleanup on `after_process_queue` is a pattern Action Scheduler's own hook invites, and it is not currently safe to write.

---

## What I have not proven

The deleted row took its own log entries with it, so I cannot read back which hook action 342695 was. By elimination it is almost certainly `wp_mail_smtp_queue_process`: line 126 requires `is_recurring()`; the action was created between 09:05:05 and 09:11:06 (bracketed by surviving `action created` logs) and executed at 09:11:06, so it was a short-interval recurrence; its logs were cleared, so it went through `delete_action()`; and the only automated same-day `delete_action()` callers on the install are the two WP Mail SMTP task hooks, of which `wp_mail_smtp_send_enqueued_email` is scheduled `->async()` (single, never recurring) and so can never reach line 126.

**This would be falsified by** finding another uncutoffed automated `delete_action()` caller, or another sub-5-minute recurring hook, on the install. I swept all plugins, mu-plugins, themes, the system crontab and its scripts for `delete_action(`, raw DELETEs against `actionscheduler_*`, `as_unschedule_*`, `after_process_queue` subscribers and the retention filters, and found none. (Note that `as_unschedule_action()` / `as_unschedule_all_actions()` are *not* candidates — `cancel_action()` is an `UPDATE` to `canceled`; unscheduling never removes a row. That eliminates ~55 call sites across WooCommerce, Subscriptions and others in one step.)

The one deleter I could not exclude on mechanism was Action Scheduler's own admin list-table Delete row action (`ActionScheduler_ListTable`), which is genuinely uncutoffed. It is now excluded on evidence: the HTTP access log covering the whole period contains **zero** requests to any Action Scheduler admin page, and no `wp-admin` request of any kind in the minutes surrounding the fatal. Nobody clicked Delete.

None of this uncertainty affects the code defect itself, which is verifiable by inspection at tag `4.1.0` independent of this site.

---

## Correction to one number, in case it is quoted

An earlier internal draft of this analysis said the hook runs "~1440×/day" on the basis of its 60-second schedule. That is wrong: `action_id` is `AUTO_INCREMENT` and each recurrence consumes exactly one id, and ids advanced 337180 → 342967 in 7.79 days — ~740/day for **every hook on the install combined**. The hook is *scheduled* every 60s; the runner does not fire that often. The retention anomaly above does not depend on the rate.
